clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+        if (it != decode_cache_.end() && it->second.is_valid()) {
+            return "";
+        }
+        decode_cache_[trace_pc] = DecodeInfo();
+        instr_t *instr = nullptr;
+        instr_noalloc_t noalloc;
+        if (persist_decoded_instrs_) {
+            instr = instr_create(dcontext_);
+        } else {
+            instr_noalloc_init(dcontext_, &noalloc);
+            instr = instr_from_noalloc(&noalloc);
+        }
+
+        app_pc decode_pc;
+        if (!use_module_mapper_) {
+            decode_pc = const_cast<app_pc>(memref_instr.encoding);
+        } else {
+            // Legacy trace support where we need the binaries.
+            std::string err = find_mapped_trace_address(trace_pc, decode_pc);
+            if (err != "")
+                return err;
+        }
+        app_pc next_pc = decode_from_copy(dcontext_, decode_pc, trace_pc, instr);
+        if (next_pc == nullptr || !instr_valid(instr)) {
+            if (persist_decoded_instrs_) {
+                instr_destroy(dcontext_, instr);
+            }
+            return "decode_from_copy failed";
+        }
+        decode_cache_[trace_pc].set_decode_info(dcontext_, memref_instr, instr);
+        return "";
+    }
+
+    /**
+     * Performs initialization tasks such as verifying whether the given trace
+     * indeed has embedded encodings or not, and initializing the
+     * \p module_mapper_t if the module path is provided.
+     *
+     * It is important to note that the trace filetype may be obtained using the
+     * get_filetype() API on a \p memtrace_stream_t. However, not all instances of
+     * \p memtrace_stream_t have the filetype available at init time (before the
+     * \p TRACE_MARKER_TYPE_FILETYPE is actually returned by the stream).
+     * Particularly, the \p stream_t provided by the \p scheduler_t to analysis
+     * tools provides the file type at init time when #dynamorio::drmemtrace::
+     * scheduler_tmpl_t::scheduler_options_t.read_inputs_in_init is set. In other
+     * cases, the user may call this API after init time when the
+     * \p TRACE_MARKER_TYPE_FILETYPE marker is seen (still before any instr
+     * record).
+     *
+     * If the provided \p module_file_path is empty, the cache object uses
+     * the encodings embedded in the trace records.
+     *
+     * If the \p module_file_path parameter is not empty, it instructs the
+     * \p decode_cache_t object that it should look for the instr encodings in the
+     * application binaries using a \p module_mapper_t. A \p module_mapper_t is
+     * instantiated only one time and reused for all objects
+     * of \p decode_cache_t (of any template type). The user must provide a valid
+     * \p module_file_path if decoding instructions from a trace that does not
+     * have embedded instruction encodings in it, as indicated by absence of
+     * the \p OFFLINE_FILE_TYPE_ENCODINGS bit in the \p TRACE_MARKER_TYPE_FILETYPE
```

Comment:
Use # for actual links for both of these.

Reply:
DOne.